Updated reuse_cmake_install with an example of package()

mristin opened the pull request:
This update is related to the issue #716 .

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/conan-io/docs?pullRequest=750) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/conan-io/docs?pullRequest=750) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/conan-io/docs?pullRequest=750) it.</sub>

danimtb commented:
Hi @mristin, thanks a lot for this PR! Could you please check the CLA sign? We need it to get this PR accepted.

Regarding the changes in this PR, one of the things that made us not kind of recommend this usage is because package() method would be called twice in case you define [no_copy_source = True](https://docs.conan.io/en/latest/reference/conanfile/attributes.html#no-copy-source).

Also, as conanfile methods are executed independently, you have to create a new CMake object in package() and configure same definitions, for example: https://github.com/bincrafters/conan-mbedtls/blob/414f0a54ba604f5fcd56b85a03874b717ae5036b/conanfile.py#L54

So finally I would say it is a trade-off. And I'd say that if we are going to make this change, it will be better to include the wholse solution as in the example linked above.

mristin commented:
@danimtb thank you for the review!

Please apologize for somewhat amateurish pull requests  -- I'm new to conan and we have just started to use it in the company. 

> Regarding the changes in this PR, one of the things that made us not kind of recommend this usage is because package() method would be called twice in case you define no_copy_source = True.

After reading the docs you pointed me to I'm a bit puzzled why the `package()` needs to be called twice. Wouldn't CMake (_i.e._ `cmake.install()`) be responsible for all the copying including the headers or any other source files? Could you please give me a bit more details so that I can also update the docs in this pull request? I suppose this merits at least a short paragraph.

> Also, as conanfile methods are executed independently, you have to create a new CMake object in package() and configure same definitions, [...]

Could you please have a look at the commit 0675a46b038118feb494ccd92d24cbd18454d28a and let me know if that is what you had in mind or something more complex?

danimtb commented:
Hi @mristin,

Don't worry, every PR (especially to the docs) is always welcomed and we really appreciate the feedback from users.

Regarding to the commit you pointed: yes, that is exactly what I was trying to highlight.

For the case of `package()` method being called twice, this happens when you have `no_copy_source = True`. First, let me explain how Conan behaves:

- Conan retrieves the sources from `exports`, `exports_sources` & `source()` and puts them in a source folder.
- Before compiling, Conan creates a copy of the source folder into the build folder
- Then the build starts.
- After the build Conan creates the package collecting the files and artifacts from the build folder (using package() method of the recipe).

Now, if you set  `no_copy_source = True` Conan behaves this way:

- Retrieve sources (same as above).
- Build starts in the build folder (sources are in source folder -NOT copied to build folder-
- After the build, Conan creates the package copying files from the source folder (this calls `package()`) and copying the artifacts from the build folder (this calls `package()` again).

So in the case you were using `cmake.install()` in the package method and `no_copy_source`, the install step will be done twice. Files will be overwritten with issues but install steps are sometimes time expensive and this will double the "packaging" time.

mristin commented:
@danimtb thanks for your explanation! 

I'm still a bit puzzled about this step:
> After the build, Conan creates the package copying files from the source folder (this calls package()) and copying the artifacts from the build folder (this calls package() again).

Why would you call the same function twice that doesn't include any parameter to change its behavior? I looked into:
https://github.com/conan-io/conan/blob/c9df68faf57003dbe480f5dbbbe887ec8aeed4a9/conans/model/conan_file.py#L247

```python
    def package(self):
        """ package the needed files from source and build folders.
        E.g. self.copy("*.h", src="src/includes", dst="includes")
        """
self.output.warn("This conanfile has no package step")
```
so there seems to be no side effects. Are there parameters I'm not aware of (like `package_folder` dynamically changing between the two steps in the concrete instance of `ConanFile`)?

mristin commented:
Hi @danimtb,
Just a kind reminder in case you missed the last message.

danimtb commented:
Hi @mristin,

As I told you before (sorry about the mistake):

> So in the case you were using `cmake.install()` in the package method and `no_copy_source`, the install step will be done twice. Files will be overwritten witho**OUT** issues but install steps are sometimes time expensive and **this will double the "packaging" time**.

Currently I thinks the best would be to make the changes you are suggesting in this PR (cmake.install() in package()) and include a warning message regarding ``no_copy_source``. For the future I think we should improve this in the conan client regarding other build helpers too.

danimtb commented on howtos/cmake_install.rst:
```diff
@@ -5,33 +5,54 @@ How to reuse cmake install for package() method
+The following excerpt from a custom script (meant to install conan, build and package the code) shows you how to set
+``package_folder`` properly:
+
+.. code-block:: python
```
Please do not include code using the Conan internal API as this is not stable enough to recommend its usage in the documentation.

danimtb commented on howtos/cmake_install.rst:
```diff
@@ -5,33 +5,54 @@ How to reuse cmake install for package() method
+        # the install prefix at this stage.
+        conan.build(cwd=build_folder, conanfile_path=source_folder, package_folder=package_folder)
 
-The ``package_info()`` method is still necessary, as there is no possible way to
```
Please, do not remove this note as it is relevant to propagate the package information and remind the user to do so

danimtb commented on howtos/cmake_install.rst:
```diff
@@ -5,33 +5,54 @@ How to reuse cmake install for package() method
+        cmake = self.configure_cmake()
+        cmake.install()
+
+If you want to specify a different ``CMAKE_INSTALL_PREFIX`` for your package, you need to specify it as
```
Dont see the point here to change CMAKE_INSTALL_PREFIX as you want it to be pointing to the package folder as it is already. Cant see the use case

mristin replied:
Please mind that we don't want to change `CMAKE_INSTALL_PREFIX`. We want to change the default `package_folder` to an arbitrary folder. Conan will pass this value to CMake which eventually sets `CMAKE_INSTALL_PREFIX`. 

We have a release script which we use to build and install in arbitrary folders. We use conan only to manage the dependencies, but don't have any downstream conan consumers.

I wanted to include this example to help other people as it took us long to figure it out, but I suppose it makes most sense if we just keep using it internally and leave it undocumented here as long as API is not stable.

mristin commented:
Hi @danimtb ,
I tried to write down what we discussed here regarding the different work flows when `no_copy_source` is set.

I also added a more verbose paragraph on `package_info()`. That part was also not clear to me, and having the redundant explanation why it is needed (aside the main `package_info()` page) would help understanding, IMO. 

I removed the example that leverages internal conan API.

memsharded commented on howtos/cmake_install.rst:
```diff
@@ -5,33 +5,69 @@ How to reuse cmake install for package() method
+    def configure_cmake(self):
         cmake = CMake(self)
         cmake.configure()
+        cmake.definitions["SOME_DEFINITION"] = "On"
```
I think this ``cmake.definitions`` should be **before** ``cmake.configure()``, otherwise it is ignored.

memsharded commented on howtos/cmake_install.rst:
```diff
@@ -5,33 +5,69 @@ How to reuse cmake install for package() method
-automatically extract the information of the necessary libraries, defines and flags for different
-build configurations from the cmake install.
+    def package_info(self):
+        self.cpp_info.includedirs = ['include']
```
These are the defaults, it is not necessary to define them, and it might be confusing to do so, without a clear warning or something.

danimtb commented:
@memsharded done! https://docs.conan.io/en/latest/howtos/cmake_install.html

- Fixed definitions.
- Changed package_id()
- Removed redundant information.
- Linked to no_copy_source reference
